src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncRewriter.AsyncIteratorRewriter.cs

```diff
@@ -100,52 +100,71 @@ protected override void GenerateControlFields()
                 _currentField = F.StateMachineField(elementType, GeneratedNames.MakeIteratorCurrentFieldName());
             }
 
-            /// <summary>
-            /// Generates the body of the replacement method, which initializes the state machine. Unlike regular async methods, we won't start it.
-            /// </summary>
-            protected override BoundStatement GenerateStateMachineCreation(LocalSymbol stateMachineVariable, NamedTypeSymbol frameType)
+            protected override void GenerateConstructor()
             {
-                // If the async method's result type is a type parameter of the method, then the AsyncTaskMethodBuilder<T>
-                // needs to use the method's type parameters inside the rewritten method body. All other methods generated
-                // during async rewriting are members of the synthesized state machine struct, and use the type parameters
-                // from the struct.
-                AsyncMethodBuilderMemberCollection methodScopeAsyncMethodBuilderMemberCollection;
-                if (!AsyncMethodBuilderMemberCollection.TryCreate(F, method, null, out methodScopeAsyncMethodBuilderMemberCollection))
+                // Produces:
+                // .ctor(int state)
+                // {
+                //     this.state = state;
+                //     this.initialThreadId = {managedThreadId};
+                //     this.builder = System.Runtime.CompilerServices.AsyncVoidMethodBuilder.Create();
+                //     this.valueOrEndPromise = new ManualResetValueTaskSourceLogic<bool>(this);
+                // }
+                Debug.Assert(stateMachineType.Constructor is IteratorConstructor);
+
+                F.CurrentFunction = stateMachineType.Constructor;
+                var bodyBuilder = ArrayBuilder<BoundStatement>.GetInstance();
+                bodyBuilder.Add(F.BaseInitialization());
+                bodyBuilder.Add(F.Assignment(F.Field(F.This(), stateField), F.Parameter(F.CurrentFunction.Parameters[0]))); // this.state = state;
+
+                var managedThreadId = MakeCurrentThreadId();
```

Comment:
As I mentioned offline, I don't really understand our use of thread id here.

The other thing to consider: if our goal is to reuse the iterators on subsequent calls, a common pattern is probably going to be,

```
getenumerator && foreach
await
getenumerator && foreach
await
...
```

During each of those awaits we could be shoved onto the thread pool and come back on a different thread, but in the same task. In those cases it seems like what we would want is not the thread ID, but an async local marking whether or not it's OK to come back.

Reply:
I added a note to [workitems list](https://github.com/dotnet/roslyn/issues/24037) to confirm the whole threadID design, to re-convince ourselves that it is sound.

Reply:
@agocke It turns out I had looked into the threadID question at some point already.
https://github.com/dotnet/corefx/issues/3481 discusses the trade-offs of this design for iterators and concluded to keep the threadID design. The same analysis applies to async-iterators.